# Allele specific RNAseq pipeline
[![Docker Build Status](https://img.shields.io/docker/automated/biocorecrg/asrnaseq.svg)](https://cloud.docker.com/u/biocorecrg/repository/docker/biocorecrg/asrnaseq/builds)
[![Nextflow version](https://img.shields.io/badge/Nextflow-19.10.0-brightgreen)](https://www.nextflow.io/)
[![Singularity version](https://img.shields.io/badge/Singularity-v2.6.1-green.svg)](https://www.sylabs.io/)
[![Docker version](https://img.shields.io/badge/Docker-v19.03-blue)](https://www.docker.com/)
<img align="right" href="https://biocore.crg.eu/" src="https://raw.githubusercontent.com/CRG-CNAG/BioCoreMiscOpen/master/logo/biocore-logo_small.png" />

This **Nextflow**[1] based workflow allows alignment of paired ends reads to a reference transcriptome described as a genome in fasta file plus an annotation in GTF format using **STAR**[2] aligner considering the variant information provided as a VCF file.
**STAR** implements the **WASP**[3] method for filtering of allele specific alignments and reports within the resulting alignments which variant is found and from which allele.
A custom python script based on **pysam** [4] efficiently separates the alignment generated by STAR into three different files:
- Reference
- Alternative
- Ambiguous (found both variants from reference and alternative on one or different pairs)

**HTseq-count**[5] tool is then used for counting tags separated in those three categories mapping to the genes.
Finally a report is generated using **multiQC**[6] that summarize the results of each step together with an initial QC evaluation fo raw reads done using **FastQC**[7]


## References
1. Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319.
1. Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner. Bioinformatics. 2013 Jan 1;29(1):15-21. 
1. van de Geijn B, McVicker G, Gilad Y, Pritchard JK. WASP: allele-specific software for robust molecular quantitative trait locus discovery. Nat Methods. 2015 Nov;12(11):1061-3.
1. https://github.com/pysam-developers/pysam
1. Ewels P, Magnusson M, Lundin S, KÃ¤ller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8.
1. https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
